### PR TITLE
Fix HF checkpointer + mlflow bugs

### PR DESCRIPTION
**Pull request by dakinggg**
Fixes an edge case where an mlflow model would not be registered if the hf save interval was exactly the same as the total duration in batches and the max duration was specified in epochs.

Manual tests for chat and not chat (and deployed and checked that input example still works), both test the edge case:
- `dur-edge-case-chat-4-mztOoR`
- `dur-edge-case-pr-3-u5qIfu`

Note, there is now a scary exception in the mlflow model saving logs, but it is harmless. The deployment and input example still works.

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -54,7 +54,8 @@
 
 install_requires = [
     'mosaicml[libcloud,wandb,oci,gcs]>=0.21.3,<0.22',
-    'mlflow>=2.10,<2.12',
+    'mlflow>=2.12.1,<2.13',
+    'pandas',
```
Comment (dakinggg) [inline setup.py:58]:
```suggestion
```

File: llmfoundry/callbacks/hf_checkpointer.py [hunk 2]
```diff
@@ -260,6 +258,13 @@ def _is_last_batch(self, state: State):
             return True
 
         assert state.max_duration is not None  # for pyright
+
```
Comment (irenedea) [inline llmfoundry/callbacks/hf_checkpointer.py:261]: Can you add a comment to describe the case that this handles?

**Review by irenedea — APPROVED** [review]

**Review by mvpatel2000 — APPROVED** [review]
Approving to unblock
